napari/components/layerlist.py

```diff
@@ -46,19 +46,31 @@ def _coerce_name(self, name, layer=None):
         new_name : str
             Coerced, unique name.
         """
-        for _layer in self:
-            if _layer is layer:
-                continue
-            if _layer.name == name:
-                name = inc_name_count(name)
-
+        if layer is None:
+            for existing_name in sorted(x.name for x in self):
+                if name == existing_name:
+                    name = inc_name_count(name)
+        else:
+            for _layer in sorted(self, key=lambda x: x.name):
+                if _layer is layer:
+                    continue
+                if name == _layer.name:
+                    name = inc_name_count(name)
         return name
 
     def _update_name(self, event):
         """Coerce name of the layer in `event.layer`."""
         layer = event.source
         layer.name = self._coerce_name(layer.name, layer)
 
+    def insert(self, index: int, value: Layer):
+        """Insert ``value`` before index."""
+        new_layer = self._type_check(value)
+        new_layer.name = self._coerce_name(new_layer.name)
+        self.events.inserting(index=index)
+        self._list.insert(index, new_layer)
+        self.events.inserted(index=index, value=new_layer)
```

Comment:
```suggestion
  super().insert(index, value)
```
Doesn't this work too and then we get the correct event emission?

Reply:
The reason I opted against that is that it will perform the `self._type_check` twice.  We need to run `_type_check` before calling `_coerce_name` (to make sure that `value` actually has a `.name` attribute)... but if we then just call `super()` it's going to check all over again

Reply:
then again... it's probably like 10µs to check the type... prefer to use super()?

Reply:
yeah, i thought that might be it. I'd prefer to use super even if `_type_check` get's called twice. The time delay is tiny and the code simplification is worth it to me

Reply:
sounds good.  committed